### PR TITLE
Separate Concat PR 

### PR DESCRIPTION
**Pull request by Chenyu-Shi**
Split the IN&CONCAT pr by deleting all IN changes

### CODE REVIEW

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 1]
```diff
@@ -702,6 +702,49 @@ class FlatbuffersExpressionEvaluator {
...
+                  std::string("tuix::Concat requires string, not ")
+                  + std::string(tuix::EnumNameFieldUnion(str->value_type()))
+                  + std::string(". You do not need to provide the data as string but the data should be serialized into string before sent to concat"));  
+              }
+              if (!str->is_null()){
+                // skipping over the null input 
+                auto str_field = static_cast<const tuix::StringField *>(str->value());
+                int32_t start = 0;
+                int32_t end = str_field ->length(); 
+                total += end;
```
Comment (wzheng) [inline src/enclave/Enclave/ExpressionEvaluation.h:729]: I think this will return a NULL value when concatenating with an empty string? Can you first add another test to test empty strings, then modify it if that's the case?

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala [hunk 2]
```diff
@@ -389,6 +389,22 @@ trait OpaqueOperatorTests extends FunSuite with BeforeAndAfterAll { self =>
     df.filter($"word".contains(lit("1"))).collect
   }
 
+  testAgainstSpark("concat with string") { securityLevel =>
```
Comment (wzheng) [inline src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:392]: Add a test case for empty string.

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 3]
```diff
@@ -702,6 +702,49 @@ class FlatbuffersExpressionEvaluator {
...
+                throw std::runtime_error(
+                  std::string("tuix::Concat requires string, not ")
+                  + std::string(tuix::EnumNameFieldUnion(str->value_type()))
+                  + std::string(". You do not need to provide the data as string but the data should be serialized into string before sent to concat"));  
+              }
+              if (!str->is_null()){
+                // skipping over the null input 
+                auto str_field = static_cast<const tuix::StringField *>(str->value());
+                int32_t start = 0;
+                int32_t end = str_field ->length();
```
Comment (wzheng) [inline src/enclave/Enclave/ExpressionEvaluation.h:728]: Why are you using int32_t here, and then casting them to uint32_t later?

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 4]
```diff
@@ -702,6 +702,49 @@ class FlatbuffersExpressionEvaluator {
...
+      std::vector<uint8_t> result; 
+
+      for (size_t i =0; i< num_children; i++){
+              auto offset =  eval_helper(row, (*c->children())[i]);
+              const tuix::Field *str = flatbuffers::GetTemporaryPointer(builder, offset);
+              if (str->value_type() != tuix::FieldUnion_StringField) {
+                throw std::runtime_error(
+                  std::string("tuix::Concat requires string, not ")
+                  + std::string(tuix::EnumNameFieldUnion(str->value_type()))
+                  + std::string(". You do not need to provide the data as string but the data should be serialized into string before sent to concat"));
```
Comment (wzheng) [inline src/enclave/Enclave/ExpressionEvaluation.h:722]: I'm a bit confused about this runtime error's message -- seems like we just do not allowing data of another type, since a runtime error is thrown?

File: src/flatbuffers/Expr.fbs [hunk 5]
```diff
@@ -118,6 +119,15 @@ table Contains {
     right:Expr;
 }
 
+// Array expressions
+table In{
+    children:[Expr]; 
+}
```
Comment (wzheng) [inline src/flatbuffers/Expr.fbs:125]: This definition shouldn't appear in this PR, since it's already in the IN PR.

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 1 above)
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h]: Good point. I added an empty string test in addition to the null test and seems like concat just Ignores both

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala (hunk 2 above)
Comment (Chenyu-Shi) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:392]: combined with "concat with other datatype test"

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 3 above)
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h]: fixed. Good point here

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 4 above)
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h]: Changed the runtime error to better reflect what needs to happen

File: src/flatbuffers/Expr.fbs (hunk 5 above)
Comment (Chenyu-Shi) [reply src/flatbuffers/Expr.fbs]: Removed
